### PR TITLE
Async handlers, beforeExit support and SIGHUP signal

### PR DESCRIPTION
**Pull request by Tapppi**
Support async handlers by inspecting whether the function signature of a handler includes an argument (the callback). Support async for beforeExit, SIGINT, SIGTERM and SIGHUP.

Added timeout to force exit after 10 seconds. Not sure what the timeout should be, or if it should be configurable though.

Async handlers originally by @mvayngrib

Also support SIGHUP as a termination signal.

### CODE REVIEW

**Comment by Tapppi** [comment]
This got way out of hand, since i forgot to branch early enough.. Also, managed to make 0.10 unsupported it seems. I pretty much rewrote the whole thing, tests (ava 0.13.0 with subprocesses to test exits) and all. You are welcome to inspect my changes at https://github.com/Tapppi/async-exit-hook

I guess I could fix the branch, but I already spent too much time on this, and have to focus on some paid work for a change :)

File: package.json [hunk 1]
```diff
@@ -28,12 +28,19 @@
...
     "event"
   ],
   "devDependencies": {
-    "ava": "0.0.4"
+    "ava": "0.13.0"
+  },
+  "ava": {
+    "files": [
+      "test/*.js",
+      "!tests/cases/*"
```
Comment (sindresorhus) [inline package.json]: Just `cases` => `fixtures` and you won't need to define any files here, as AVA ignores `fixtures` folders.

**Comment by sindresorhus** [comment]
Awesome! Thanks for working on this. I'll try to get to review this on monday ;)

**Comment by Tapppi** [comment]
Sure, I'll try to put together a cleaner version tomorrow if I have the time. The Master branch in my fork is way cleaner.

File: test/tests.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	return new Promise(resolve => {
+		var proc = fork(
+			path.resolve(__dirname, './cases/' + test + '.js'),
+			{
+				env: process.env,
+				silent: true
+			}
+		);
+
+		var output = '';
```
Comment (SamVerschueren) [inline test/tests.js]: `let`

File: test/tests.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * exit code and output in an array ([code, output])
+ *
+ * @async
+ * @param {String} test Filename without path or extension
+ * @param {String} signal Signal (or 'shutdown' for message) to send to the process
+ * @return {[Number, String]} Array with the exit code and output of the subprocess
+ */
+function testInSub(test, signal) {
+	return new Promise(resolve => {
+		var proc = fork(
```
Comment (SamVerschueren) [inline test/tests.js]: `const`

File: test/tests.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Tests have to happen in a subprocess to test the exit functionality
+'use strict';
+
+var test = require('ava');
+var fork = require('child_process').fork;
+var path = require('path');
+
```
Comment (SamVerschueren) [inline test/tests.js]:
```js
import path from 'path';
import {fork} from 'child_process';
import test from 'ava';
```

File: test/tests.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+test('sync handlers', t => {
+	t.plan(2);
+	return testInSub('sync', 'shutdown')
+		.then(([code, output]) => {
+			t.is(output, 'SUCCESS');
+			t.is(code, 0);
+		});
+});
```
Comment (SamVerschueren) [inline test/tests.js]:
```js
test('sync handlers', async t => {
    const [output, code] = await testInSub('sync', 'shutdown');
    t.is(output, 'SUCCESS');
    t.is(code, 0);
});
```

Other tests should be using the same structure with `async/await`.

**Comment by beefchimi** [comment]
Any progress on this? `async` is exactly what I need 😄

**Pull request closed without merging**